박스오피스 [STEP 4] Sunny, Blu

calledBlu opened the pull request:
>안녕하세요! 븨븨(@YebinKim)!
🧴**TEAM SUNBLock**🧴 의 Sunny, Blu입니다!
저번 븨븨가 용산에 놀러 오셨을 때, STEP3-2에 대한 대략적인 리뷰를 받기도 했고(대면 리뷰를 받을 수 있음에 행복했어요.😀)
촉박하게 프로젝트를 구현했기 때문에 이번 PR은 STEP 3-2와 STEP 4를 같이 보내게 되었어요🥰
>
>이번 리뷰도 잘 부탁드립니다!💙☀️
>븨븨의 리뷰를 받을 수 있어서 정말 알차고 즐거웠던 프로젝트였어요! 👍👍

<br>

# 🧑🏼‍💻 작업 목록
## 네트워크 통신 및 영화 상세화면 구현
- [x] 이미지 검색 API와의 통신
    - API Key를 Header에 포함하여 통신
    - 이미지 검색 결과를 받아올 모델타입 구현
    - 네트워크 통신을 담당하는 타입 개선
- [x] 영화 상세정보를 나타내는 뷰 구현
    - StackView 활용
    - Safe Area을 고려한 오토 레이아웃 구현
    - 나타내야 하는 정보
        - 영화 포스터 이미지
        - 영화 상세정보
- [x] refresh control 개선

<br>

# 실행 화면


<br>

# 📚 학습 키워드 및 학습 내용
## 1. Protocol과 Generic의 차이

- **학습 내용**
   - Generic: 정적인 다형성 구현 가능
       - 타입이 한번 지정되면 제네릭으로 지정된 모든 타입이 한번에 지정됨
       - generic이 protocol 내에 존재하는 경우 원래는 타입으로 사용할 수 없음!
       - 이를 가능하도록 하는 것이 `any`와 `some`키워드
   - Protocol: 동적인 다형성 구현 가능
       - Protocol을 채택하는 다양한 타입이 들어갈 수 있음
- **적용 과정**
   - `PresentationModel`(현 `PresentationProvider`)를 구현할 당시 endPoint의 다양한 EndPoint를 수용하기 위해 프로토콜을 사용하려고 함
        ```swift
        struct PresentationModel: PresentationProvidable {

            private let networkProvider: NetworkProvidable
            private var date: String {
                return Date.yesterday.formatData(type: .network)
            }

            private var endpoint: RequestResponseProtocol

            init(networkProvider: NetworkProvidable = NetworkProvider()) {
                self.networkProvider = networkProvider
                self.endpoint = DailyBoxOfficeEndpoint(date: date)
            }

            func dispatch(endpoint: RequestResponseProtocol) async {

                // NetworkProvider를 통해 데이터를 받아오는 request 함수 실행
                // networkProvider.request(endpoint)
            }
        }
        ```
- **문제와 해결**
    - 😰 위 코드 중 `private var endpoint: RequestResponseProtocol` 부분에서 `RequestResponseProtocol`을 타입으로 사용할 수 없어서 문제 발생
    - `RequestResponseProtocol`은 `Requestable & Responsable` 을 합성한 프로토콜이며 `Responsable` 프로토콜에서는 `associatedtype`을 가지고 있음!
    - 🚀 protocol 내에 associatedtype이 존재하므로 타입으로 사용할 수 없으므로 `Endpoint`라는 Generic을 생성하여 해당 Generic이 `RequestResponseProtocol`을 채택하도록 코드 수정
        ```swift
        struct PresentationModel<Endpoint: RequestResponseProtocol>: PresentationProvidable {

            private let networkProvider: NetworkProvidable
            private var date: String {

                return Date.yesterday.formatData(type: .network)
            }

            private var endpoint: Endpoint

            init(networkProvider: NetworkProvidable = NetworkProvider(), endpoint: Endpoint) {
                self.networkProvider = networkProvider
                self.endpoint = endpoint
            }
        ```
- **아쉬운 점**
   - endPoint를 내부에 가지고 있는 방향으로 코드를 수정하여 해결한 코드를 사용하지 못하게 된 것

<br>

## 2. UI 레이아웃 구성 방법

- **학습 내용**
    - iOS와 macOS의 사용자 인터페이스의 레이아웃은 3가지 방법으로 구성 가능
    - Frame-based Layout
        - view의 위치와 크기 정보를 frame으로 표현하는 것
    - Autoresizing Masks
        - Super view의 frame이 변경될 때 view의 frame이 변경되는 방식
        - 늘어난 Super view의 크기에 맞춰 자연스럽게 화면이 배치됨    
        예) `collectionView.autoresizingMask = [.flexibleWidth, .flexibleHeight]`
    - Auto Layout
        - view의 크기, 위치, view 사이 간격에 constraint를 설정하여 조정
        - view의 frame 대신 view 사이의 관계를 정의하여 내부/외부 변화에 모두 대응함
    - 적용되는 순서는`Autoresizing Masks` -> `autoLayout` -> `Frame`
- **적용 과정**
    - collection view의 레이아웃을 `Autoresizing Masks`로 조정하려고 함
- **문제 해결**
    - 😰 Autoresizing Masks를 사용하였을 때 view의 safeArea를 벗어나는 문제 발생
    - 🚀 Autoresizing Masks 대신 Auto Layout을 사용하는 방향으로 변경

<br>

## 3. UICalendarView

> [참고문서 - UICalendarView](https://developer.apple.com/documentation/uikit/uicalendarview)
- **학습 내용**
    - `UICalendarView`는 iOS 16.0 이상부터 지원하기 때문에 버전을 상향시켜야 함
    - 이전에는 UIDatePicker로 캘린더 뷰를 구성하였음
    - `UICalendarSelectionSingleDate`를 통해 선택한 날짜를 활용할 수 있음
- **적용 과정**
    - 핵심 경험으로 `UICalendarView`의 사용이 있었기 때문에 최소 지원버전을 16.0으로 상향하기로 결정
    - `dateSelection(_:,didSelectDate:)`에서 calendar 클로저 프로퍼티를 호출하여 PresentationProvider의 날짜를 update
    - 프로젝트 요구사항에는 오늘 날짜까지 선택할 수 있도록 하였으나, 오늘 날짜를 선택하는 경우 박스오피스 데이터는 존재하지 않기 때문에 어제 날짜까지만 선택할 수 있도록 정정
- **문제 해결**
    - 😰 날짜 선택 시, 실제 선택한 날짜와 LLDB로 출력한 날짜가 하루정도 차이가 남
    - 🚀 메서드를 통해 selected 된 날짜는 UTC+0을 기준으로 하지만, 한국의 경우 UTC+9를 적용해야 함, 이를 적용하기 위해 Calendar 타입의 `date(bySettingHour:, minute:, second:, of:`메서드를 활용 bySettingHour 매개변수로 9시간을 전달하여 날짜를 일치하게 함
- **아쉬운 점**
    - 선택한 날짜가 미리 선택되어 있는 부분을 아직 구현하지 못해서 아쉽지만 추후 학습을 통해 구현 예정

<br>

# 👾 구현 세부 사항
## 1️⃣ 영화 포스터를 Daum API를 통해 가져오기

#### 1. `Network`: DaumAPI, MoviePosterDTO 및 MoviePosterEndpoint 생성
- 영화 포스터 이미지를 가져오기 위해 DaumAPI를 사용
- MovieAPI와 다른 점으로는 API key가 headers에 들어가 있음

#### 2. `PresentationModel`: MovieInformationItem 타입에 프로퍼티 추가
- 영화 포스터를 위한 PresentationModel을 따로 생성하지 않고, 영화 세부사항을 나타내는 MovieInformationItem 타입 안에 poster 프로퍼티 추가

#### 3. `Presentation - Dispatcher`: MovieInformationDispatcher 타입에 함수 추가
- 영화 포스터를 위한 Dispatcher을 따로 생성하지 않았음
- 영화 포스터 또한 MovieInformationItem 의 하나이기 때문에 해당 Dispatcher에 영화 포스터를 가져오는 함수를 추가
- `fetchMoviePosterDTO` : `MoviePosterEndpoint`  ➡️ `MoviePosterDTO` 로 변환하는 함수
- `convertImage` : `MoviePosterDTO`로 받아온 이미지URL을 `UIImage` 로 변환하는 함수

#### 4. `Presentation - Provider`: PresentationProvider 타입에 함수 추가
- 영화의 세부 정보를 load하는 함수 내부에 영화 포스터를 load하는 함수 호출
- `PresentationProvider` 내부에 영화 세부정보 호출을 위한 movieCode 프로퍼티 생성
    - didSet을 사용해 movieCode의 값이 변경될 경우, 변경된 영화 코드로 새로운 영화 세부정보 업데이트
    ```swift
   private var movieCode: String = "" {
        didSet {
            loadMovieInformation()
        }
    }
    ```
- `loadMovieInformation` : movieCode를 사용해 영화 세부정보를 가져오는 함수
    ```swift
    func loadMovieInformation() {

        let movieInformationEndpoint = MovieInformationEndpoint(movieCode: self.movieCode)

        Task {
            let networkData = try await self.movieInformationDispatcher.fetch(endpoint: movieInformationEndpoint)
            guard var movieInformation = try self.movieInformationDispatcher.convert(from: networkData).first else { return }

            movieInformation.poster = try await loadMoviePoster(movieName: movieInformation.movieName)
            self.movieInformation = movieInformation

            movieInformationCall?()
        }
    }
    ```
    - `movieInformationCall`을 사용해 movieInformation load가 모두 완료되면 완료된 시점 전달

- `loadMoviePoster` : 영화 이름으로 영화 포스터를 가져오는 함수
    ```swift
        func loadMoviePoster(movieName: String) async throws -> UIImage? {

        let moviePosterEndpoint = MoviePosterEndpoint(movieName: movieName)

        let networkData = try await self.movieInformationDispatcher.fetchMoviePosterDTO(moviePosterEndpoint)
     
        let moviePoster = try self.movieInformationDispatcher.convertImage(from: networkData)

        return moviePoster
    }
    ```

## 2️⃣ BoxOfficeViewController에서 PresentationProvider을 사용해 데이터를 로드 및 화면 업데이트

- BoxOfficeViewController의 `viewDidLoad()` 내부의 `setupInitialBoxOffices` 함수 실행
    ```swift
    private func setupInitialBoxOffices() {
        
        presentationProvider.loadBoxOffices()
        configureDataSource {
            self.indicatorView.stopAnimating()
        }
    }
    ```
    - BoxOfficeViewController에 보여질 BoxOffices 초기 데이터를 설정하는 함수
    - ViewController는 데이터를 갖지 않는다. 모든 데이터는 `PresentationProvider`가 갖고 있고, ViewController는 해당 타입의 함수들을 호출할 뿐이다.
- CollectionView의 `DataSource` 설정
     ```swift
    private func configureDataSource(completionHandler: @escaping (() -> Void)) {

        presentationProvider.boxOfficeCall = {
            let boxOfficeDataSource = BoxOfficeDataSource()
            boxOfficeDataSource.boxOffices = self.boxOffices
            self.dataSource = boxOfficeDataSource

            DispatchQueue.main.async {
                self.title = self.presentationProvider.getBoxOfficeDate()
                self.collectionView.dataSource = self.dataSource
                completionHandler()
            }
        }
    }
    ```
    - `presentationProvider.loadBoxOffices()` 호출이 완료되면, `boxOfficeCall`이 실행된다.
    - BoxOffices 데이터 로드가 완료되었기 때문에, collectionView의 dataSource에 새로운 데이터를 갖고 있는 `BoxOfficeDataSource()`를 넣어준다.
    - `presentationProvider.getBoxOfficeDate()` 를 호출해 title을 변경한다.

<br>

> 왜 BoxOfficeViewController가 박스오피스를 조회하는 `날짜` 프로퍼티를 갖지 않는가?

- Controller는 상태에 따라 변화하는 값을 갖지 않는다. 
- 이러한 이유 때문에 박스오피스 호출을 위한 `일자` 속성을 `PresentationProvider`가 갖는다.
- `PresentationProvider`의 `date, movieCode`는 `private` 로 선언되었기 때문에, 데이터를 가져오거나 업데이트 할 때 직접 접근하는 것이 아니라 함수를 이용한다.
```swift
func getBoxOffices() -> [BoxOfficeItem] {
    return self.boxOffices
}

func getBoxOfficeDate() -> String {
    return self.date.formatData(type: .title)
}

func getMovieInformation() -> MovieInformationItem? {
    return movieInformation
}

func updateDate(_ date: Date) {
    self.date = date
}

func updateMovieCode(_ movieCode: String) {
    self.movieCode = movieCode
}
```

<br>

# 🔫 트러블 슈팅
## 영화 포스터 이미지 비율대로 설정하기

### 문제 발생

<img src="https://hackmd.io/_uploads/ByY4hh-S2.png" width="260" height="450">

- 영화 세부정보를 보여주는 화면에서 영화 포스터 자체의 이미지 비율이 아니라 StackView의 layout에 꽉 차서 이미지가 나오는 문제 발생
- 이미지 고유의 비율이 아니기 때문에 이미지가 부자연스럽게 보이는 현상이 발생

### 문제 도전
아래 키워드에 대해 학습하면서 이미지 크기 조정을 시도함

- **contentMode** : View의 bounds가 변경될 때, content가 어떻게 배치될지 결정하는데 사용하는 mode

- contentMode를 쉽게 이해할 수 있는 그림 [출처](https://blog.naver.com/PostView.naver?blogId=soojin_2604&logNo=222439470655&parentCategoryNo=&categoryNo=&viewDate=&isShowPopularPosts=false&from=postView)
- <img src="https://hackmd.io/_uploads/SkCGah-S2.png" width="400" height="500"> 
    
- contentMode.scaleAspectFit : content 고유의 비율은 유지하면서, View 크기에 맞게 확장, 나머지 영역은 투명하게 변경됨

### 문제 해결
<img src="https://hackmd.io/_uploads/HkWw03bH2.png" width="260" height="450">

- UIImageView에 `contentMode` 설정
    ```swift
    private let moviePoster: UIImageView = {

        let imageView = UIImageView()
        imageView.heightAnchor.constraint(equalToConstant: 400).isActive = true
        imageView.contentMode = .scaleAspectFit
        return imageView
    }()
    ```


<br>

SunnnySong committed:
feat: PresentationProvider 생성

- 추상부인 PresentationProvidable 생성
- 구현부인 PresentationProvider 생성

calledBlu committed:
refactor: View에 데이터를 보여줄 수 있도록 ViewController 수정

- `presentationProvider` 타입을 ViewController가 가지고 있도록 구현
- `presentationProvider`의 delegate를 `BoxOfficeViewController`로 지정
- `presentationProvider`를 초기화할 경우 호출되는 delegate의 `call()` 메서드를 통해 데이터를 업데이트하도록 구현
- 위의 코드 수정에 따른 `BoxOfficeDataSource` 코드 수정

calledBlu committed:
feat: String을 확장하여 관객 수를 세 자리 수마다 `,`로 구분하도록 구현

- String을 확장하여 String에서 Int 타입으로 변환하는 프로퍼티와 세 자리 수마다 `,`로 구분하는 `formatted()` 메서드 구현
- `BoxOfficeItem`의 이니셜라이저를 활용하여 초기화 시 audience를 변환된 숫자를 포함한 값으로 저장되도록 구현
- 위의 수정에 따른 `BoxOfficeDispatcher` 코드 수정
- mock 데이터 삭제 및 불필요한 공백 제거

SunnnySong committed:
feat: cell의 disclosureIndicator 추가

SunnnySong committed:
refactor: PresentationProvider에서 boxOfficeCall 수정

- 기존의 PresentationDelegate 삭제
- boxOfficeCall 을 생성해서, boxOffice가 생성되면 클로저 호출

calledBlu committed:
feat: Daum 이미지 검색을 사용하기 위한 API 타입 구현

- `DaumAPI` 타입 생성 및 API에 필요한 속성 구현
- 이미지 검색 API를 사용하기 위한 Key 추가 및 Bundle에서 경로를 가져와서 사용할 수 있도록 구현
- 새로운 API key 추가에 따른 기존 영화 관련 API Key 네이밍 변경
- 영화 상세정보 DTO들이 채택한 `Codable`을 `Decodable`로 변경

SunnnySong committed:
feat: MoviePosterDTO / Endpoint 생성

calledBlu committed:
feat: MovieInformationItem 타입 구현

SunnnySong committed:
feat: MovieInformationDispatcher 생성

- 영화의 세부 정보를 가져오는 함수를 가진 MovieInformationDispatcher 생성
- MovieInformationItem의 poster 프로퍼티 옵셔널 변경
- PresentationDispatchableㅇ에서 convert 함수 async 제

calledBlu committed:
refactor: 상세정보와 포스터를 가져오도록 PresentationProvider 수정

- `date`, `movieCode` 프로퍼티를 `private`로 전환
- 상세정보와 포스터를 가져오는 메서드 추가
- 위의 수정사항에 따른 BoxOfficeViewController 수정
- refresh control이 데이터 로딩 중에는 스크롤을 멈추도록 수정

SunnnySong committed:
refactor: BoxOfficeViewController BoxOffices 로드 코드 수정

- setupInitialBoxOffices 함수 생성 : 초기 BoxOffices 데이터 설정 및

calledBlu committed:
feat: 영화 셀을 탭하면 상세정보로 이동하도록 구현

- BoxOfficeViewController에서 셀을 탭하는 경우 해당 영화의 상세정보를 보여주는 Controller와 연결되도록 구현
- 영화 코드를 업데이트하는 `updateMovieCode(_:)`메서드 생성
- `BoxOfficeItem`이 영화 코드를 가지고 있도록 수정 및 해당 수정에 따른 `BoxOfficeDispatcher` 수정

SunnnySong committed:
feat: MovieInformationView 생성

- MovieInformationStackView 생성
- MovieInformationView 생성 및 MovieInformationViewController와 연
- PresentationProvider에서 클로저 데이터 전달 구현

calledBlu committed:
chore: DTO 속성 네이밍 변경 및 영화 상세정보 뷰 indicator 추가

SunnnySong committed:
Auto stash before merge of "STEP4" and "origin/STEP4"

SunnnySong committed:
feat: MovieInformationView 에서 ScrollView 추가

- scrollView, contentView 추가 및 autoLayout 설정

calledBlu committed:
feat: 날짜 선택을 위해 `UICalendarView`를 통한 캘린더 구현

- `CalendarViewController` 생성
- `UICalendarView` 사용을 위해 최소 버전 14.0 -> 16.0으로 상향 조정
- 날짜 선택 시 해당 날짜를 반환하고 반환한 날짜의 박스오피스 목록을 가져오도록 구현
- 날짜 선택을 어제 날짜로 제한(오늘의 박스오피스는 불러올 수 없기 때문!)

SunnnySong committed:
Revert "feat: CalendarView 추가"

This reverts commit d075a3886578f5d91877bf31725940cca866a914.

SunnnySong committed:
feat: CalendarView 추가

- CalendarViewController 생성
    - calendarView 생성 및 속성 정의
- BoxOfficeViewController에서 UIBarButtonItem 추가

SunnnySong committed:
Auto stash before revert of "feat: CalendarView 추가"

SunnnySong committed:
Merge remote-tracking branch 'origin/STEP4' into STEP4

SunnnySong committed:
chore: 줄 바꿈 및 코드 정리

calledBlu committed:
refactor: Calendar 날짜 선택 시 날짜가 일치하지 않던 문제 해결

YebinKim commented on BoxOffice/Controller/ViewController/MovieInformationViewController.swift:
```diff
@@ -0,0 +1,57 @@
+            DispatchQueue.main.async {
+                self.indicatorView.stopAnimating()
+            }
+        }
```
p1;

메모리 누수가 발생하고 있네요! 수정이 필요해보입니다 👀

SunnnySong replied:
[weak self] 를 추가해 약한 참조로 메모리 누수를 해결했어요!@ 
deinit을 추가해 확인해보니 메모리 누수를 쉽게 알 수 있어서 편리했어요😀
꿀팁 감사드려요!

YebinKim commented on BoxOffice/Controller/ViewController/BoxOfficeViewController.swift:
```diff
@@ -0,0 +1,123 @@
+        presentationProvider.loadBoxOffices()
+        configureDataSource {
+            self.indicatorView.stopAnimating()
+        }
```
p1;

마찬가지로 메모리 누수가 발생하고 있어요! `BoxOfficeViewController` 파일 내에서는 총 세 군데 발생합니다 👀

SunnnySong replied:
수정 완료했습니다😉

YebinKim commented on BoxOffice/View/MovieInformation/MovieInformationStackView.swift:
```diff
@@ -0,0 +1,50 @@
+
+import UIKit
+
+class MovieInformationStackView: UIStackView {
```
p3;

상속 가능성이 없는 class 는 final 처리를 해두면 빌드 성능 향상에 도움이 되어요! 😊
일부 클래스만 final 처리가 되지 않은 거 같아서 확인 부탁드려요.

calledBlu replied:
final 키워드가 필요한 클래스들을 찾아 추가해 주었어요!

> 해당 커밋: 4d94932

YebinKim commented on BoxOffice/View/MovieInformation/MovieInformationView.swift:
```diff
@@ -0,0 +1,97 @@
+        didSet {
+            configureMovieInformation()
+        }
+    }
```
p2;

`information`은 외부에서 접근하고 변경 또한 가능한 상태 값으로 보이네요 👀
현재 사용하는 로직을 살펴보면 `MovieInformationViewController` 가 세팅될 때(`viewDidLoad()`에서) 설정되는 값으로 보이는데, `configureMovieInformation` 메서드를 통해 매개변수 형태로 받아오는 형태로 만들어보면 어떨까요?

calledBlu replied:
`information` 프로퍼티 캡슐화 및 이를 위한 메서드를 구현해 보았습니다!

> 해당 커밋: 32d5158

YebinKim commented on BoxOffice/View/BoxOfficeRank/BoxOfficeDataSource.swift:
```diff
@@ -8,16 +8,18 @@
 
 final class BoxOfficeDataSource: NSObject, UICollectionViewDataSource {
+
+    var boxOffices: [BoxOfficeItem] = []
```
p3;

`boxOffices` 는 언제든 외부에서 접근 가능하고 변경 가능하기 때문에 캡슐화가 잘 되어 있지 않은 것으로 보여요 🤔
생성자를 통해 주입받는 형태로 변경해보면 어떨까요?

SunnnySong replied:
`boxOffices`에 private 키워드를 붙여 접근 제어를 설정했어요.
대신 븨븨의 코멘트를 받아 생성자를 통해 해당 프로퍼티 값을 설정해 줄 수 있도록 수정했어요😀

YebinKim commented on BoxOffice/Controller/ViewController/MovieInformationViewController.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        view.addSubview(indicatorView)
```
p2;

super 메서드의 호출이 누락되어 있네요 👀
메서드가 제대로 동작할 수 있도록 super 메서드의 호출이 필요해보여요!

관련 문서: [viewWillAppear(_:)](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621510-viewwillappear)
> If you override this method, you must call super at some point in your implementation.

calledBlu replied:
관련 문서를 보면서 다시 한번 복습 후 수정하였습니다!

> 해당 커밋: 4d94932

YebinKim reviewed (changes requested):
써니, 블루 안녕하세요~
박스오피스 Step4 진행까지 고생 많으셨어요 🥳🎉
이번에도 타입의 추상화 구조에 신경을 많이 쓰셨음이 느껴졌어요 👍
Protocol 의 연관값 문법을 멋지게 활용하신 구조 또한 눈에 띄었습니다!! 🤩

또한, 데이터 전달 및 동작 전달을 위해 컴플리션 구조를 주로 활용하신 걸로 보이는데요 👀
이로 인해 발생하는 메모리 누수가 일부 있어서 코멘트 남겨두었으니 확인 부탁드려요~

---

이번 스텝의 핵심 경험은 아래와 같았는데요.
<img width="308" alt="step4" src="https://github.com/tasty-code/ios-box-office/assets/38066039/4174f8c1-fae1-4a37-b86c-c7698962d3f1">

스텝 핵심 경험과는 별개로 p1, p2 코멘트가 있어서 Request Changes 로 요청드렸어요.

뷰컨트롤러의 deinit 에 로그를 달아두면 뷰컨트롤러 내부에서 메모리 누수가 발생하는지 확인하기 편해요! 제가 주로 사용하는 방법입니다 😆
```swift
deinit {
	debugPrint("deinit \(self)")
}
```

SunnnySong committed:
refactor: 메모리 누수 수정'

- BoxOfficeViewController [weak self] 키워드 추가
- MovieInformationViewController [weak self] 키워드 추가

SunnnySong committed:
refactor: 메모리 누수 수정'

- BoxOfficeViewController [weak self] 키워드 추가
- MovieInformationViewController [weak self] 키워드 추가

SunnnySong committed:
Merge remote-tracking branch 'origin/STEP4' into STEP4

calledBlu committed:
refactor: final 키워드 추가 및 super 메서드 호출

- `MovieInformationViewController`, `MovieInformationStackView` class에 final 키워드 추가
- `MovieInformationViewController`의 `viewWillAppear(_:)` 메서드의 super 메서드 호출

SunnnySong committed:
refactor: BoxOfficeDataSource 내부 boxOffices private 설정

- BoxOfficeDataSource의 프로퍼티인 boxOffices 속성 private 키워드 추가
- 생성자를 통해 boxOffices 받아오게 수정

calledBlu committed:
refactor: MovieInformationView의 일부 속성 캡슐화

- information을 외부에서 접근할 수 없도록 private 키워드 추가
- information 속성을 업데이트하는 `updateInformation(information: )` 메서드 구현

SunnnySong committed:
chore: MovieInformationView init()에서 불필요한 코드 삭제

YebinKim reviewed (approved):
써니, 블루 Step4 수정 커밋 모두 확인했습니다~
이번 스텝 및 프로젝트는 여기서 깔끔하게 마무리하게 되었네요! 약 4주 간 박스오피스 프로젝트 진행하느라 정말 고생 많으셨어요 ☺️
